src/v/serde/avro/parser.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "container/fragmented_vector.h"
+
+#include <avro/Schema.hh>
+
+#include <variant>
+
+namespace serde::avro {
+
+namespace parsed {
+
+struct record;
+struct list;
+struct map;
+struct avro_null {};
+struct avro_union;
+/**
+ * Avro primitive being either a number, enum, string, bytes, avro fixed or null
+ */
+using primitive = std::variant<
+  avro_null,
+  double,
+  float,
+  int32_t, // Can be an enum value
+  int64_t,
+  bool,
+  iobuf // iobuf can represent either bytes, string or fixed types
+  >;
+
+/**
+ * Variant representing a result of Avro message deserialization
+ */
+using message = std::variant<primitive, map, record, avro_union, list>;
+
+/**
+ * Avro map represented as a vector of pairs, the same as in Avro. Avro MAP does
+ * not guarantee unique keys
+ */
+struct map {
+    chunked_vector<std::pair<iobuf, std::unique_ptr<message>>> entries;
+};
+
+struct list {
+    chunked_vector<std::unique_ptr<message>> elements;
+};
+
+/**
+ * Avro record, representing a struct with a vector of fields. As for the AVRO
+ * specification the order of fields in the vector is the exactly the same as
+ * the order of fields in the schema, AVRO do not support optional fields. The
+ * optionals are often expressed as a union of an actual type and AVRO_NULL,
+ * which is expressed as a separate entry in the fields vector.
+ */
+struct record {
+    chunked_vector<std::unique_ptr<message>> fields;
+};
+
+/**
+ * AVRO union is decoded in a special way to make it possible to easily corelate
+ * the parsing result with Avro schema.
+ */
+struct avro_union {
+    // an index of the type that the union is holding
+    size_t branch;
+    std::unique_ptr<message> message;
```

Comment:
I think this is the only place in the current type structure where a unique_ptr is required. It's probably best to be consistent before trying to optimize anything for performance. There might be a structure where no unique_ptr is needed (as I don't know if it's valid avro to have a union of a union), but against let's not over optimize until we have real perf numbers.